docs/okeps/okep-5094-layer2-transit-router.md

```diff
@@ -0,0 +1,981 @@
+Currently, every Layer3 topology (the most switches/routers) needs 7 keys (node switch, cluster router, join switch, 
+gw router, transit router, transit switch, ext switch), so we could reserve 20 keys per network (just to be safe).
+These keys could be derived from the network-id, and then we could support e.g. 100K UDNs (2M keys) + 1M UDN interconnects (1M keys),
+and still have 12M keys left.
+
+The easiest way for now is to use option 1. We can always introduce the option 3 later, preserving already-assigned keys.
+We should add error handling in case we run out of tunnel keys both for transit routers and transit switches.
+
+#### NAT configuration
+As much as possible everything related to conntrack should not be modified since doing so can affect tcp connections.
+
+The only nat rule that needs to be moved from GR to transit_router is the one SNATing the traffic from the network subnet that goes to the mgmt port. This is needed because now ovn_custer_router is the router directly connected to the layer2 switch (instead of the GR). The rest of the NAT configurations can stay unchanged on the GR, not also that
+there are plans to move the management port SNAT from OVN to iptables, so there were no issue here.
+
+```
+allowed_ext_ips     : []
+exempted_ext_ips    : []
+external_ids        : {"k8s.ovn.org/network"=test12_namespace-scoped, "k8s.ovn.org/topology"=layer2}
+external_ip         : "169.254.0.12"
+external_mac        : []
+external_port_range : ""
+gateway_port        : []
+logical_ip          : "203.203.0.0/16"
+logical_port        : rtos-test12_namespace.scoped_ovn_layer2_switch
+match               : "eth.dst == 0a:58:cb:cb:00:02"
+options             : {stateless="false"}
+priority            : 0
+type                : snat
+```
+
+Also the fact that join IP at gateway router is kept allow to maintain all the NATing done with it at gateway router to define OVN load balancers to implement k8s services.
+
+#### Static Routes and Logical Route Policies
+Changing where routes and policies are configured does not affect TCP connections on upgrade so we can move some routes and policies from gateway router
+to transit_router without issues. In general it will be similar to the layer3 routes and policies.
+
+These will be the routes and policies configured on the new transit_router
+```
+IPv4 Routes
+Route Table <main>:
+                10.96.0.1               203.203.0.2 dst-ip <-- api service
+               10.96.0.10               203.203.0.2 dst-ip <-- dns service
+               100.65.0.2                100.88.0.6 dst-ip <-- node1 traffic use node2 gr peer ip
+               100.65.0.3               100.88.0.10 dst-ip <-- node2 traffic use node2 gr peer ip
+               100.65.0.4               100.88.0.14 dst-ip <-- node3 traffic use node2 gr peer ip
+           203.203.0.0/16               100.88.0.14 src-ip <-- network egress traffic goes to local gateway router peer ip
```

Comment:
hmm i forget what this is used for? was it eips?

Reply:
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/56d14a3074ba7df00a6b68dfe4ac60aff7fc624a/go-controller/pkg/libovsdb/util/router.go#L26